### PR TITLE
pd: report hot read cpu in heartbeat

### LINKED ISSUE #19373: pd: report hot read cpu in heartbeat
ref https://github.com/tikv/pd/issues/5718

### PR DESCRIPTION
**Pull request by lhy1024**
### What is changed and how it works?

Issue Number: Close https://github.com/tikv/tikv/issues/19373

What's Changed:

```commit-message
 Improve hotspot read accounting by reporting per-region unified-read CPU in `cpu_stats`, preventing CPU stat loss across region/store heartbeat paths, and fixing Linux thread-name truncation
  matching for read CPU attribution.
- Add `cpu_stats` propagation on the Region heartbeat reporting path (`RegionStat` -> PD request).
  - Keep `cpu_usage` as a compatibility field (deprecated in kvproto) for rolling upgrades.
  - Split region CPU accumulators into two maps:
    - `since_region_heartbeat`
    - `since_store_heartbeat`
    so one heartbeat path no longer drains CPU data required by the other.
  - Include unified-read CPU in Store heartbeat peer_stats, and add CPU into hotspot filtering and TopN selection.
    TopN dimensions are expanded from 3 to 4 (keys/bytes/query/cpu).
  - Do not consume peer deltas / CPU records on fake store heartbeats, preserving full intervals for the next real heartbeat.
  - Clean up both CPU maps when peers are destroyed, and clear orphan CPU records.
  - Add prefix matching utility for Linux 15-char thread-name truncation to fix read CPU attribution for long prefixes (for example `unified-read-pool`).
  - Add unit tests for CPU thresholding, TopN collection, zero-interval handling, map cleanup, and thread-name prefix matching.
  - Update kvproto revision (with cpu_stats-related fields), and add a temporary denylist entry for `RUSTSEC-2026-0009` in `deny.toml` due to current `time` upgrade constraints with raft-engine/serde
  pinning.
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test

### Release note

```release-note
None
```

## Summary by CodeRabbit

* **New Features**
  * CPU statistics are now attached to region and store heartbeats, used for hotspot detection, Top-N CPU reporting, and gated reporting.

* **Chores**
  * Dependency for kvproto pinned to a specific revision for reproducible builds.
  * Improved thread-name prefix matching for CPU aggregation.
  * Per-region CPU accounting split to support distinct heartbeat/reporting flows.

* **Tests**
  * Unit tests added for per-region state removal, CPU-threshold reporting, and thread-name prefix behavior.

### CODE REVIEW

File: components/tikv_util/src/thread_name_prefix.rs [hunk 1]
```diff
@@ -154,3 +154,43 @@ pub const SNAP_SENDER_THREAD: &str = "snap-sender";
 pub const TABLET_SNAP_SENDER_THREAD: &str = "tablet-snap";
 
 pub const STATUS_SERVER_THREAD: &str = "status-server";
+
+const LINUX_THREAD_NAME_MAX_LEN: usize = 15;
+
+#[inline]
+pub fn matches_thread_name_prefix(thread_name: &str, prefix: &str) -> bool {
```
Comment (okJiang) [inline components/tikv_util/src/thread_name_prefix.rs:170]: Please add more comments to this function

File: components/raftstore/src/store/worker/pd.rs [hunk 2]
```diff
@@ -1274,36 +1297,56 @@ where
...
+                    report_peers.insert(*region_id, read_stat);
                 }
-                let mut read_stat = pdpb::PeerStat::default();
-                read_stat.set_region_id(*region_id);
-                read_stat.set_read_keys(read_keys);
-                read_stat.set_read_bytes(read_bytes);
-                read_stat.set_query_stats(query_stats.0);
-                report_peers.insert(*region_id, read_stat);
             }
+            stats = collect_report_read_peer_stats(HOTSPOT_REPORT_CAPACITY, report_peers, stats);
```
Comment (okJiang) [inline components/raftstore/src/store/worker/pd.rs]: How about wrap these to a function?

File: components/raftstore/src/store/worker/pd.rs [hunk 3]
```diff
@@ -952,6 +972,8 @@ where
 
     // region_id -> total_cpu_time_ms (since last region heartbeat)
     region_cpu_records: HashMap<u64, u32>,
+    // region_id -> total_cpu_time_ms (since last store heartbeat)
+    region_cpu_records_store: HashMap<u64, u32>,
```
Comment (okJiang) [inline components/raftstore/src/store/worker/pd.rs]:
```suggestion
// region_id -> total_cpu_time_ms (since last region heartbeat)
region_cpu_records_since_region_heartbeat: HashMap<u64, u32>,
    // region_id -> total_cpu_time_ms (since last store heartbeat)
    region_cpu_records_since_store_heartbeat: HashMap<u64, u32>,
```

File: components/raftstore-v2/src/worker/pd/mod.rs [hunk 4]
```diff
@@ -217,6 +217,8 @@ where
     region_buckets: HashMap<u64, region::ReportBucket>,
     // region_id -> total_cpu_time_ms (since last region heartbeat)
     region_cpu_records: HashMap<u64, u32>,
+    // region_id -> total_cpu_time_ms (since last store heartbeat)
+    region_cpu_records_store: HashMap<u64, u32>,
```
Comment (rleungx) [inline components/raftstore-v2/src/worker/pd/mod.rs]: What is the difference between `region_cpu_records` and `region_cpu_records_store`?

File: components/raftstore/src/store/worker/pd.rs [hunk 5]
```diff
@@ -924,6 +924,75 @@ fn hotspot_query_num_report_threshold() -> u64 {
     HOTSPOT_QUERY_RATE_THRESHOLD * 10
 }
 
+fn hotspot_cpu_usage_report_threshold() -> u64 {
+    const HOTSPOT_CPU_USAGE_THRESHOLD: u64 = 1;
```
Comment (rleungx) [inline components/raftstore/src/store/worker/pd.rs]: Why not move it to line 907?

File: components/raftstore/src/store/worker/pd.rs [hunk 6]
```diff
@@ -924,6 +924,75 @@ fn hotspot_query_num_report_threshold() -> u64 {
...
+) -> bool {
+    !(read_bytes < hotspot_byte_report_threshold()
+        && read_keys < hotspot_key_report_threshold()
+        && query_stats.get_read_query_num() < hotspot_query_num_report_threshold()
+        && cpu_usage < hotspot_cpu_usage_report_threshold())
+}
+
+fn collect_report_peers_for_store_heartbeat(
+    region_peers: &mut HashMap<u64, PeerStat>,
+    region_cpu_records_since_store_heartbeat: &mut HashMap<u64, u32>,
```
Comment (rleungx) [inline components/raftstore/src/store/worker/pd.rs:951]: Does it mean since last?

**Review by okJiang — APPROVED** [review]

File: components/raftstore/src/store/worker/pd.rs [hunk 7]
```diff
@@ -924,6 +931,71 @@ fn hotspot_query_num_report_threshold() -> u64 {
...
+        let query_stats = region_peer
+            .query_stats
+            .sub_query_stats(&region_peer.last_store_report_query_stats);
+        let cpu_usage = if interval_seconds > 0 {
+            let cpu_time_duration = Duration::from_millis(
+                region_cpu_records_since_store_heartbeat
+                    .remove(region_id)
+                    .unwrap_or(0) as u64,
+            );
+            ((cpu_time_duration.as_secs_f64() * 100.0) / interval_seconds as f64) as u64
```
Comment (LykxSassinator) [inline components/raftstore/src/store/worker/pd.rs:968]: Although the case `interval_seconds == 0` is not expected under normal operating conditions, it should should be explicitly handled for safety.

File: components/raftstore/src/store/worker/pd.rs [hunk 8]
```diff
@@ -2649,7 +2735,7 @@ fn collect_report_read_peer_stats(
     mut report_read_stats: HashMap<u64, pdpb::PeerStat>,
     mut stats: pdpb::StoreStats,
 ) -> pdpb::StoreStats {
-    if report_read_stats.len() < capacity * 3 {
+    if report_read_stats.len() < capacity * 4 {
```
Comment (LykxSassinator) [inline components/raftstore/src/store/worker/pd.rs]: Is this modification expected ? Should we enlarge it ?

File: Cargo.toml [hunk 9]
```diff
@@ -375,7 +375,7 @@ grpcio-health = { version = "0.10.4", default-features = false, features = [
   "protobuf-codec",
 ] }
 tipb = { git = "https://github.com/pingcap/tipb.git" }
-kvproto = { git = "https://github.com/pingcap/kvproto.git" }
+kvproto = { git = "https://github.com/pingcap/kvproto.git", rev = "c5ab144dc6c7b4eb2091270b3f06e8ccc1b47c8a" }
```
Comment (LykxSassinator) [inline Cargo.toml]: Please remove "rev" after all relevant commits in `kvproto` are merged

File: components/raftstore/src/store/worker/pd.rs (hunk 8 above)
Comment (lhy1024) [reply components/raftstore/src/store/worker/pd.rs]: yes, because we add cpu in statistics.

File: components/raftstore/src/store/worker/pd.rs [hunk 10]
```diff
@@ -924,6 +932,74 @@ fn hotspot_query_num_report_threshold() -> u64 {
     HOTSPOT_QUERY_RATE_THRESHOLD * 10
 }
 
+#[inline]
+fn should_report_read_peer(
```
Comment (LykxSassinator) [inline components/raftstore/src/store/worker/pd.rs]:
```suggestion
fn should_report_hotspot_read_peer(
```

**Review by LykxSassinator — COMMENTED** [review]
Rest LGTM

**Review by LykxSassinator — APPROVED** [review]
LGTM

**Review by rleungx — APPROVED** [review]
